feat: Enhance apply_diff with XML for multi-file/multi-diff operations & batch UI

samhvw8 opened the pull request:
<!--
Thank you for contributing to Roo Code!

Before submitting your PR, please ensure:
- It's linked to an approved GitHub Issue.
- You've reviewed our [Contributing Guidelines](../../CONTRIBUTING.md).
-->

### Related GitHub Issue
This pr will fix this issues https://discord.com/channels/1332146336664915968/1367897808249819157


<!-- Every PR MUST be linked to an approved issue. -->

Closes: #2362 <!-- Replace with the issue number, e.g., Closes: #123 -->

### Description
Convert to xml tag for all params and support multiple write file
<!--
Briefly summarize the changes in this PR and how they address the linked issue.
The issue should cover the "what" and "why"; this section should focus on:
- The "how": key implementation details, design choices, or trade-offs made.
- Anything specific reviewers should pay attention to in this PR.
-->

### Test Procedure

<!--
Detail the steps to test your changes. This helps reviewers verify your work.
- How did you test this specific implementation? (e.g., unit tests, manual testing steps)
- How can reviewers reproduce your tests or verify the fix/feature?
- Include relevant testing environment details if applicable.
-->

### Type of Change

<!-- Mark all applicable boxes with an 'x'. -->

- [x] 🐛 **Bug Fix**: Non-breaking change that fixes an issue.
- [x] ✨ **New Feature**: Non-breaking change that adds functionality.
- [ ] 💥 **Breaking Change**: Fix or feature that would cause existing functionality to not work as expected.
- [ ] ♻️ **Refactor**: Code change that neither fixes a bug nor adds a feature.
- [ ] 💅 **Style**: Changes that do not affect the meaning of the code (white-space, formatting, etc.).
- [ ] 📚 **Documentation**: Updates to documentation files.
- [ ] ⚙️ **Build/CI**: Changes to the build process or CI configuration.
- [ ] 🧹 **Chore**: Other changes that don't modify `src` or test files.

### Pre-Submission Checklist

<!-- Go through this checklist before marking your PR as ready for review. -->

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [ ] **Code Quality**:
    - [ ] My code adheres to the project's style guidelines.
    - [ ] There are no new linting errors or warnings (`npm run lint`).
    - [ ] All debug code (e.g., `console.log`) has been removed.
- [ ] **Testing**:
    - [ ] New and/or updated tests have been added to cover my changes.
    - [ ] All tests pass locally (`npm test`).
    - [ ] The application builds successfully with my changes.
- [ ] **Branch Hygiene**: My branch is up-to-date (rebased) with the `main` branch.
- [ ] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [ ] **Changeset**: A changeset has been created using `npm run changeset` if this PR includes user-facing changes or dependency updates.
- [ ] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](../../CONTRIBUTING.md).

### Screenshots / Videos

<!--
For UI changes, please provide before-and-after screenshots or a short video of the *actual results*.
This greatly helps in understanding the visual impact of your changes.
-->

### Documentation Updates

<!--
Does this PR necessitate updates to user-facing documentation?
- [ ] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).
-->

### Additional Notes

<!-- Add any other context, questions, or information for reviewers here. -->

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Enhance `apply_diff` with XML support for multi-file/multi-diff operations and introduce batch UI for improved user interaction.
> 
>   - **Behavior**:
>     - `apply_diff` function now supports XML for multi-file/multi-diff operations in `multiApplyDiffTool.ts`.
>     - Batch UI for multi-file operations added in `BatchDiffApproval.tsx`.
>     - Legacy `applyDiffTool` renamed to `applyDiffToolLegacy`.
>   - **Experiments**:
>     - New experiment `MULTI_FILE_APPLY_DIFF` added to `experiments.ts` and related files.
>     - Dynamic strategy selection based on experiment state in `Task.ts` and `generateSystemPrompt.ts`.
>   - **Tests**:
>     - Added tests for new multi-file diff functionality in `applyDiffTool.experiment.spec.ts` and `Task.test.ts`.
>   - **UI**:
>     - Added batch diff approval UI component in `BatchDiffApproval.tsx`.
>     - Updated i18n files for new batch operation messages.
>   - **Misc**:
>     - Minor refactoring and renaming for clarity and consistency.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for b7e3202f06a9454ea59403a2c88dbfbb3335f7b4. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

samhvw8 commented:
example of test

[roo_task_may-9-2025_12-45-30-am.md](https://github.com/user-attachments/files/20107497/roo_task_may-9-2025_12-45-30-am.md)

KJ7LNW commented:
Neat!  Can writes be confirmed individually?

samhvw8 commented:
> Neat!  Can writes be confirmed individually?

Yes, it confirm in each file ( i group diff into each file )

KJ7LNW commented:
This 3342 PR is going to conflict heavily with #3055

any ideas on how to prevent that?

samhvw8 commented:
> This 3342 PR is going to conflict heavily with #3055
> 
> any ideas on how to prevent that?

i will try make a branch merge conflict for you

samhvw8 commented:
@KJ7LNW  i make this backward compatible too 💪

samhvw8 commented:
wait this merge pr 
 https://github.com/RooVetGit/Roo-Code/pull/3654

samhvw8 commented:
@daniel-lxs  @hannesrudolph  @KJ7LNW  can you recheck this ui 

following UI from multiple_read_file

<img width="706" alt="image" src="https://github.com/user-attachments/assets/90e67209-0bc0-4850-9b7f-bfffb8592bae" />

iamhenry commented:
right. happy to test other themes. at the moment i'm using the light vercel theme but it shouldnt matter. reusing the component i think is the most important takeaway.

daniel-lxs reviewed (commented):
Hey @samhvw8, this looks pretty good, just a question about error messages sent to the model.

Also would it make sense to add this new functionality to the Experimental settings just like the multi file read feature?

It might also be worth addressing some of the points @ellipsis-dev left.

samhvw8 commented:
> Hey @samhvw8, this looks pretty good, just a question about error messages sent to the model.
> 
> 
> 
> Also would it make sense to add this new functionality to the Experimental settings just like the multi file read feature?
> 
> 
> 
> It might also be worth addressing some of the points @ellipsis-dev left.

Let me think more about this, if you have ideas can you share about it

KJ7LNW commented on src/core/tools/applyDiffTool.ts:
@hannesrudolph  cleaned up `read_file` some time ago so that all responses to the model from read file are organized as XML.

something like this could be useful and it would clean up the response:

https://github.com/KJ7LNW/cline/blob/c91d5642bbb20afe7a825424c50df969bfd55f6f/src/integrations/editor/DiffViewProvider.ts#L285-L326

samhvw8 commented:
update exp setting, @daniel-lxs  can you tuning prompt for me :(

samhvw8 commented:
@ellipsis review this pr again for me

daniel-lxs commented:
Converting to draft so it's not accidentally merged, as discussed with @samhvw8, the tool instructions expose the new XML format even when the experiment flag is disabled. This teaches models to use a format that's only partially functional, leading to confusion.

We need to hide the XML examples/format from the tool description when `MULTI_FILE_APPLY_DIFF` is false, showing only the legacy format until the experiment is enabled.

samhvw8 commented:
@daniel-lxs pushed update but we need to test it again 🤣

daniel-lxs reviewed (approved):
LGTM

daniel-lxs commented on src/core/tools/multiApplyDiffTool.ts:
```diff
@@ -0,0 +1,571 @@
+			throw new Error(detailedError)
+		}
+
+		// Remove this duplicate check - we already checked at the beginning
```
```suggestion
```

samhvw8 commented:
@daniel-lxs  Thank You So Much <3 <3
